AsyncDisplayKit/Private/ASDisplayNode+UIViewBridge.mm

```diff
@@ -282,212 +288,222 @@ - (void)setNeedsDisplay
       [rasterizedContainerNode setNeedsDisplay];
     });
   } else {
-    // If not rasterized (and therefore we certainly have a view or layer),
-    // Send the message to the view/layer first, as scheduleNodeForDisplay may call -displayIfNeeded.
-    // Wrapped / synchronous nodes created with initWithView/LayerBlock: do not need scheduleNodeForDisplay,
-    // as they don't need to display in the working range at all - since at all times onscreen, one
-    // -setNeedsDisplay to the CALayer will result in a synchronous display in the next frame.
-
-    _messageToViewOrLayer(setNeedsDisplay);
-
-    BOOL nowDisplay = ASInterfaceStateIncludesDisplay(_interfaceState);
-    // FIXME: This should not need to recursively display, so create a non-recursive variant.
-    // The semantics of setNeedsDisplay (as defined by CALayer behavior) are not recursive.
-    if (_layer && !_flags.synchronous && nowDisplay && [self __implementsDisplay]) {
-      [ASDisplayNode scheduleNodeForRecursiveDisplay:self];
+    if (self.nodeLoaded) {
+      if (ASDisplayNodeThreadIsMain()) {
+        _messageToViewOrLayer(setNeedsDisplay);
+      } else {
+        if (!_pendingViewState.hasChanges) {
+          [ASPendingStateController.sharedInstance registerNode:self];
+        }
+        [self __setNeedsDisplay];
+        [_pendingViewState setNeedsDisplay];
+      }
+    } else {
+      [self __setNeedsDisplay];
     }
   }
 }
 
 - (void)setNeedsLayout
 {
-  _bridge_prologue;
-  [self __setNeedsLayout];
-  _messageToViewOrLayer(setNeedsLayout);
+  _bridge_prologue_write;
+  if (self.nodeLoaded) {
+    if (ASDisplayNodeThreadIsMain()) {
+       [self __setNeedsLayout];
+        _messageToViewOrLayer(setNeedsLayout);
+    } else {
+      if (!_pendingViewState.hasChanges) {
+        [ASPendingStateController.sharedInstance registerNode:self];
+      }
+      // NOTE: We will call [self __setNeedsLayout] just before we apply
```

Comment:
Why should this be done before the application of the pending state? It seems like during the measurement pass, some of the values in the pending state may be useful. I suppose the reading path should still provide access to those if the pending state exists and has not been cleared?